awslabs/homomorphic-implementors-toolkit
# Backend API Refactor

*Issue #, if available:*

*Description of changes:*
The primary goal of this PR is to create an internal API for the HE backend (SEAL) so that evaluators mostly call an inner API rather than SEAL directly. The purpose of this new API is to facilitate adding a different HE backend to HIT. It's a bit difficult to see the effects of this refactor in this PR alone, but these changes dramatically reduce the number of files touched by changing to the Lattigo backend. 

For example, now instead of using SEAL-specific functions inside the tests, tests use the `HEContext` API, and therefore don't need to be updated when the backend is changed. Similarly, most of the logging statements in `DebugEvaluator` that print HE parameters used to rely on SEAL-specific APIs; now they rely only on the generic `HEContext` APIs, and therefore don't need to be changed when moving to Lattigo.

Backend-specific functions are now isolated to
- `HEContext`
- `Ciphertext`
- `HomomorphicEvaluator`
- `DebugEvaluator` (minor)
- `ScaleEstimator` (minor)

Specific changes:
- Updated to latest SEAL release
- Move functions from `sealutils` into the new `HEContext` API, which is intended to capture backend-specific functionality.
- Remove unused imports
- Update a few miscellaneous comments, fix typos, and rename a few variable.
- Made a few `Evaluator` API functions _not_ `const` anymore. This is unfortunate, but necessary since the functions can't be `const` with Lattigo.
- Use SEAL `EncryptionParameter` serialization. Unclear why I wasn't using this before, but it simplifies instance serialization and makes the process more similar to how it works with Lattigo.
- Fix `enable-clang-format.cmake` script to properly report when clang-format isn't found.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
